### DISCUSSION TITLE
Regex on custom image tag

### DISCUSSION THREAD

**Discussion by karlschriek** [renovatebot/renovate#14585] [Request Help]
I am trying to devise a versioning regex that would allow Renovate to update image tags that look like this:

```
release-1.4-l0-release-1.4.1
release-1.4-l0-release-1.4.1-rc1-64-gb859213b7
release-1.4-l0-release-1.4
release-1.4-l0-release-1.4-rc2-54-ge8fe0275b
release-1.4-l0-release-1.4-rc2-51-gac4309324
release-1.4-l0-release-1.4-rc1-47-g5caff5e03
```

1.4.1 is the current latest release. Before that was 1.4.1-rc1, before that was 1.4 etc.

In addition, there will be further release branches, such as 1.2, 1.3 etc each of which I would need to update separately. So here I have to exactly match `release-1.4-l0-release-`, then `major.minor.patch`, match _anything_ on the rc part and then update build number (51 -> 54 for example) and again allow anything on the final hash.

So it needs to be something like:

`"regex:^release-1.4-l0-release-(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)\\-(?<compatibility>-)(?<build>\\d+)\\-(?<compatibility>)$"`

I feel like I am getting myself all tied up in knots though. The regex doesn't seem to be valid.

**Comment by viceice** [comment]
Check the docs, you need named capture groups for the parts

**Reply by karlschriek** [reply]
Sorry, message auto-formatting stripped out the capture groups for some reason. Fixed it now, should be `"regex:^release-1.4-l0-release-(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)\\-(?<compatibility>-)(?<build>\\d+)\\-(?<compatibility>)$"`, not "regex:^release-1.4-l0-release-(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)\\-(?<compatibility>-)(?<build>\\d+)\\-(?<compatibility>)$" as it originally showed

**Reply by viceice** [reply]
You can't use a capture group twice.

BTW: Why the version is twice in that tag? Wouldn't it easier to only create one of that?

**Reply by karlschriek** [reply]
The "release-1.4-l0-release-" part you can view as a branch name. The use-case I am working on is for publishing docs for several release branches. So for each branch we'll need to do the versioning separately. The trickiest part seems to be figuring out how to have renovate recognise that `1.4` is _newer_ than `1.4-rc2-54-ge8fe0275b`.

**Reply by karlschriek** [reply]
The closest I've managed to get so far is this:

`"regex:^release-1.4-l0-release-(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?(\\-rc(?s).*?)?(\\-(?<build>\\d+))?(\\-.*)?$`

This will work correct for transitioning between RCs, but it will not do `1.4-rc2-54-ge8fe0275b` -> `1.4`, for that I need to somehow make it recognise that the rc2 part (or even the build part for that matter) is a pre-release version. According to this (https://github.com/renovatebot/renovate/blob/main/docs/usage/docker.md#version-compatibility), there is a `prerelease` capture group as well, but I can't find any more information on it.

I've tried using this, which is actually accepted as a valid regex, but it will still not do `1.4-rc2-54-ge8fe0275b` -> `1.4`

`"regex:^something-(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?(\\-rc(?<prerelease>\\d+))?(\\-(?<build>\\d+))?(\\-.*)?$"`

**Reply by viceice** [reply]
You need to include the RC part in the prerelease group

**Reply by karlschriek** [reply]
I think you mean like this, right?

`"regex:^something-(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?(\\-(?<prerelease>\\d+))?(\\-(?<build>\\d+))?(\\-.*)?$"`

Could you maybe explain why? It is a bit hard to to extrapolate what is happening here.

(By the way this still doesn't work, possibly because the "build" part is throwing it off). I guess what I am looking for is something like

`release-1.4-l0-release-(major).(minor).(patch)-(prerelease)` where the `prerelease` part is the whole of `rc2-54` or even just `54` since that is the really relevant number here.

**Reply by viceice** [reply]
I would try something like

```json
"regex:^release-1.4-l0-release-(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?(\\-(?<prerelease>.+))?$"
```

**Reply by karlschriek** [reply]
Yes! I am going to do a few more tests, but it looks like that solves it!

**Reply by karlschriek** [reply]
So after some testing, I found out that the `(\\-(?<prerelease>.+))` will in fact not work correctly. For example

`1.4-rc2-8-xxxxxx -> 1.4-rc2-9-xxxxx` This will work correctly

But, this will not `1.4-rc2-9-xxxxxx -> 1.4-rc2-10-xxxxx` since it thinks that `9` comes after `1`
